host/traceh.c

```diff
@@ -67,88 +71,152 @@ void initialize_log_config()
+                    stderr,
+                    "%s\n",
+                    "Custom log format does not end with a newline");
+                exit(1);
+            }
+        }
         _initialized = true;
     }
 }
 
-static void _write_header_info_to_stream(FILE* stream)
+static void _escape_characters(
+    char* log_msg,
+    char* log_msg_escaped,
+    size_t msg_size)
 {
-    time_t t = time(NULL);
-    struct tm* lt = localtime(&t);
+    size_t idx = 0;
 
-    fprintf(
-        stream, "================= New logging session =================\n");
-    fprintf(stream, "%s", asctime(lt));
-    fprintf(
-        stream,
-        "https://github.com/openenclave/openenclave branch:%s\n",
-        OE_REPO_BRANCH_NAME);
-    fprintf(stream, "Last commit:%s\n\n", OE_REPO_LAST_COMMIT);
+    for (size_t i = 0; i < msg_size; i++)
+    {
+        int c = log_msg[i];
+        if (log_msg[i] == '\"')
+        {
+            // single quotes are OK for JSON
+            log_msg_escaped[idx] = (char)putchar('\'');
+        }
+        else if (log_msg[i] == '\\')
+        {
+            log_msg_escaped[idx] = '\\';
+            idx++;
+            log_msg_escaped[idx] = '\\';
+        }
+        else if (!isprint(c))
+        {
+            log_msg_escaped[idx] = '\\';
+            idx++;
+            log_msg_escaped[idx] = '\\';
+            idx++;
+            switch (log_msg[i])
+            {
+                case '\a':
+                    log_msg_escaped[idx] = 'a';
+                    break;
+                case '\b':
+                    log_msg_escaped[idx] = 'b';
+                    break;
+                case '\e':
+                    log_msg_escaped[idx] = 'e';
+                    break;
+                case '\f':
+                    log_msg_escaped[idx] = 'f';
+                    break;
+                case '\n':
+                    log_msg_escaped[idx] = 'n';
+                    break;
+                case '\r':
+                    log_msg_escaped[idx] = 'r';
+                    break;
+                case '\t':
+                    log_msg_escaped[idx] = 't';
+                    break;
+                case '\v':
+                    log_msg_escaped[idx] = 'v';
+                    break;
+                case '\\':
+                    log_msg_escaped[idx] = '\\';
+                    break;
+                case '\'':
+                    log_msg_escaped[idx] = '\'';
+                    break;
+                case '\"':
+                    log_msg_escaped[idx] = '\"';
+                    break;
+                case '\?':
+                    log_msg_escaped[idx] = '?';
+                    break;
+                default:
+                    sprintf(&log_msg_escaped[idx], "u%04x", log_msg[i]);
+                    idx += 3;
+                    break;
+            }
+        }
+        else
+        {
+            log_msg_escaped[idx] = log_msg[i];
+        }
+        idx++;
+    }
+    if (idx < 2 * msg_size)
```

Comment:
That's a hardcoded asumption about the length of log_msg_escaped, which may or may not hold?